src/ApiHandler.ts

```diff
@@ -124,20 +130,79 @@ export default class ApiHandler {
 			}
 		}
 
+		const perByteFee = api.consts.transactionPayment.transactionByteFee;
+		const extrinsicBaseWeight = api.consts.system.extrinsicBaseWeight;
+		const coefficients = api.consts.transactionPayment.weightToFee[0];
+		const coeffFrac = coefficients.coeffFrac;
+		const coeffInt = coefficients.coeffInteger;
+
+		const multiplier = (await api.query.transactionPayment.nextFeeMultiplier.at(parentHash)).div(new BN(1e9)).div(new BN(1e9));
+		
+		// TODO: We assume that the polynomial has one positive term of degree one.
+		// const degree = coefficients.degree;
+		// const negative = coefficients.negative;
+
 		for (let idx = 0; idx < block.extrinsics.length; ++idx) {
 			if (!extrinsics[idx].paysFee || !block.extrinsics[idx].isSigned) {
 				continue;
 			}
 
 			try {
-				// This is only a temporary solution. This runtime RPC will not work if types or logic
-				// involved in fee calculation changes in a runtime upgrade. For a long-term solution,
-				// we need to calculate fees based on the metadata constants and fee multiplier.
-				// https://github.com/paritytech/substrate-api-sidecar/issues/45
-				extrinsics[idx].info = api.createType(
-					'RuntimeDispatchInfo',
-					await api.rpc.payment.queryInfo(block.extrinsics[idx].toHex(), parentHash)
-				);
+				if (api.runtimeVersion.specName.toString() === 'kusama') {
+					extrinsics[idx].info = api.createType(
+						'RuntimeDispatchInfo',
+						await api.rpc.payment.queryInfo(block.extrinsics[idx].toHex(), parentHash)
+					);
+
+					continue;
+				}
+
+				const xtEvents = extrinsics[idx].events;
+				const completedEvent = xtEvents.find((event) => event.method === successEvent || event.method === failureEvent);
+				if (!completedEvent) {
+					extrinsics[idx].info = {
+						error: 'Unable to find success or failure event for extrinsic'
+					};
+
+					continue;
+				}
+
+				const completedData = completedEvent.data;
+				if (!completedData) {
+					extrinsics[idx].info = {
+						error: 'Success or failure event for extrinsic does not contain expected data'
+					};
+
+					continue;
+				}
+
+				// both ExtrinsicSuccess and ExtrinsicFailed events have DispatchInfo types as their
+				// final arg
+				const weightInfo = completedData[completedData.length - 1] as DispatchInfo;
+				if (!weightInfo.weight) {
+					extrinsics[idx].info = {
+						error: 'Success or failure event for extrinsic does not specify weight'
+					};
+
+					continue;
+				}
+
+				const weight = weightInfo.weight;
+				const encodedLength = block.extrinsics[idx].encodedLength;
+
+				const weightedCoeffFrac = weight.mul(coeffFrac).div(new BN(1e9));
```

Comment:
`div(new BN(1e9)) === divn(1e9)`, it's shorter, and I think also more performant.

Reply:
Unfortunately `bn.js` does not like this :tired_face: